src/Temporalio/Bridge/CustomSlotSupplier.cs

```
@@ -0,0 +1,179 @@
+        private readonly Dictionary<uint, Temporalio.Worker.Tuning.ISlotPermit> permits = new();
+        private uint permitId = 1;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CustomSlotSupplier" /> class.
+        /// </summary>
+        /// <param name="userSupplier">User's slot supplier implementation'.</param>
+        /// <param name="loggerFactory">Logger factory.</param>
+        internal unsafe CustomSlotSupplier(
+            Temporalio.Worker.Tuning.ICustomSlotSupplier userSupplier,
+            ILoggerFactory loggerFactory)
+        {
+            this.logger = loggerFactory.CreateLogger<CustomSlotSupplier>();
+            this.userSupplier = userSupplier;
+
+            var interopCallbacks = new Interop.CustomSlotSupplierCallbacks
+            {
+                reserve = FunctionPointer<Interop.CustomReserveSlotCallback>(Reserve),
+                cancel_reserve = FunctionPointer<Interop.CustomCancelReserveCallback>(CancelReserve),
+                try_reserve = FunctionPointer<Interop.CustomTryReserveSlotCallback>(TryReserve),
+                mark_used = FunctionPointer<Interop.CustomMarkSlotUsedCallback>(MarkUsed),
+                release = FunctionPointer<Interop.CustomReleaseSlotCallback>(Release),
+            };
+
+            PinCallbackHolder(interopCallbacks);
+        }
+
+        private static void SetCancelTokenOnCtx(ref SlotReserveCtx ctx, CancellationTokenSource cancelTokenSrc)
+        {
+            unsafe
+            {
+                try
+                {
+                    var handle = GCHandle.Alloc(cancelTokenSrc);
+                    fixed (Interop.SlotReserveCtx* p = &ctx)
+                    {
+                        Interop.Methods.set_reserve_cancel_target(p, GCHandle.ToIntPtr(handle).ToPointer());
+                    }
+                }
+                catch (Exception e)
+                {
+                    Console.WriteLine($"Error setting cancel token on ctx: {e}");
```

```suggestion
```

I assume leftover debug code? Can just remove the whole try/catch and let it bubble

---

Yeah, leftover

---

When Rust panics I have seen that converted to `System.Runtime.InteropServices.SEHException`, but that's about it